### PR TITLE
New additions and changes

### PR DESCRIPTION
**Pull request by Qudix**

### CODE REVIEW

File: include/RE/N/NiPoint3.h [hunk 1]
```diff
@@ -28,6 +28,9 @@ namespace RE
 		NiPoint3&	 operator*=(float a_scalar);
 		NiPoint3&	 operator/=(float a_scalar);
 
+		static float GetDistance(const NiPoint3& a_lhs, const NiPoint3& a_rhs);
+		static float GetSquaredDistance(const NiPoint3& a_lhs, const NiPoint3& a_rhs);
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.h:32]: Couldn't these be member functions?

File: include/RE/T/TESObjectCELL.cpp [hunk 2]
```diff
@@ -67,6 +67,35 @@ namespace RE
 	}
 
 
+	void TESObjectCELL::ForEachReference(std::function<bool(RE::TESObjectREFR* a_ref)> a_fn) const
+	{
+		spinLock.Lock();
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TESObjectCELL.cpp:72]: Please use a smart locker for this

File: include/RE/T/TESObjectCELL.cpp [hunk 3]
```diff
@@ -67,6 +67,35 @@ namespace RE
...
+		for (auto& refPtr : references) {
+			auto ref = refPtr.get();
+			if (ref) {
+				if (!a_fn(ref)) {
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TESObjectCELL.cpp:76]:
```suggestion
		for (const auto& ref : references) {
			if (ref) {
				if (!a_fn(ref.get())) {
```

File: include/RE/T/TESObjectCELL.cpp [hunk 4]
```diff
@@ -67,6 +67,35 @@ namespace RE
...
+			}
+		}
+		spinLock.Unlock();
+	}
+
+
+	void TESObjectCELL::ForEachReferenceInRange(const NiPoint3& a_origin, float a_radius, std::function<bool(TESObjectREFR* a_ref)> a_fn) const
+	{
+		ForEachReference([&](TESObjectREFR* ref) {
+			auto distance = NiPoint3::GetSquaredDistance(a_origin, ref->GetPosition());
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TESObjectCELL.cpp:88]:
```suggestion
			const auto distance = NiPoint3::GetSquaredDistance(a_origin, ref->GetPosition());
```

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

**Comment by Qudix** [comment]
Anything else?

File: include/RE/B/BSFurnitureMarkerNode.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	class BSFurnitureMarkerNode : public NiExtraData
+	{
+	public:
+		inline static constexpr auto RTTI = RTTI_BSFurnitureMarkerNode;
+		inline static constexpr auto NiRTTI = NiRTTI_BSFurnitureMarkerNode;
+
+		BSTArray<FurniturePosition> positions;	// 18
+	};
+}
```
[the remark this answers is not in the document]
Comment (Ryan-rsm-McKenzie) [reply include/RE/B/BSFurnitureMarkerNode.h:45]: you're still missing a newline

File: include/RE/T/TESObjectCELL.h [hunk 6]
```diff
@@ -204,6 +204,8 @@ namespace RE
 		INTERIOR_DATA* GetLighting();
 		float		   GetNorthRotation();
 		TESForm*	   GetOwner();
+		void		   ForEachReference(std::function<bool(TESObjectREFR* a_ref)> a_fn) const;
+		void		   ForEachReferenceInRange(const NiPoint3& a_origin, float a_radius, std::function<bool(TESObjectREFR* a_ref)> a_fn) const;
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TESObjectCELL.h:208]: Surely these callbacks should be passed references instead of pointers if they're never null?

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

File: include/RE/T/TESObjectCELL.h (hunk 6 above)
Comment (Qudix) [reply include/RE/T/TESObjectCELL.h]: Is it important?

Comment (Qudix) [reply include/RE/T/TESObjectCELL.h]: I'll just try it

File: include/RE/G/GridCellArray.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		virtual void Unk_06(void) override;	 // 06
+		virtual void Unk_07(void) override;	 // 07
+		virtual void Unk_08(void) override;	 // 08
+
+		TESObjectCELL* GetCell(std::uint32_t a_x, std::uint32_t a_y)
+		{
+			if (a_x >= length || a_y >= length) {
+				return nullptr;
+			}
+			return cells[(a_x * length) + a_y];
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/G/GridCellArray.h:32]:
```suggestion
			return (a_x < length && a_y < length) ?
				cells[(a_x * length) + a_y] :
				nullptr;
```

File: include/RE/G/GridCellArray.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+		// override (GridArray)
+		virtual void Unk_02(void) override;	 // 02
+		virtual void Unk_03(void) override;	 // 03
+		virtual void Unk_05(void) override;	 // 05
+		virtual void Unk_06(void) override;	 // 06
+		virtual void Unk_07(void) override;	 // 07
+		virtual void Unk_08(void) override;	 // 08
+
+		TESObjectCELL* GetCell(std::uint32_t a_x, std::uint32_t a_y)
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/G/GridCellArray.h:27]:
```suggestion
		[[nodiscard]] TESObjectCELL* GetCell(std::uint32_t a_x, std::uint32_t a_y) const noexcept
```

File: include/RE/N/NiPoint3.h [hunk 9]
```diff
@@ -28,6 +28,8 @@ namespace RE
 		NiPoint3&	 operator*=(float a_scalar);
 		NiPoint3&	 operator/=(float a_scalar);
 
+		float	 GetDistance(const NiPoint3& a_pt) const;
+		float	 GetSquaredDistance(const NiPoint3& a_pt) const;
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.h:32]:
```suggestion
		[[nodiscard]] float	 GetDistance(const NiPoint3& a_pt) const noexcept;
		[[nodiscard]] float	 GetSquaredDistance(const NiPoint3& a_pt) const noexcept;
```

File: include/RE/N/NiPoint3.cpp [hunk 10]
```diff
@@ -98,6 +98,24 @@ namespace RE
 	}
 
 
+	float NiPoint3::GetDistance(const NiPoint3& a_pt) const
+	{
+		float dx = a_pt.x - x;
+		float dy = a_pt.y - y;
+		float dz = a_pt.z - z;
+		return std::sqrtf(dx * dx + dy * dy + dz * dz);
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.cpp:106]:
```suggestion
		return std::sqrtf(GetSquaredDistance(a_pt));
```

File: include/RE/N/NiPoint3.cpp [hunk 11]
```diff
@@ -98,6 +98,24 @@ namespace RE
...
+	}
+
+
+	float NiPoint3::GetSquaredDistance(const NiPoint3& a_pt) const
+	{
+		float dx = a_pt.x - x;
+		float dy = a_pt.y - y;
+		float dz = a_pt.z - z;
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.cpp:114]:
```suggestion
		const float dx = a_pt.x - x;
		const float dy = a_pt.y - y;
		const float dz = a_pt.z - z;
```

File: include/RE/N/NiPoint3.cpp [hunk 12]
```diff
@@ -98,6 +98,24 @@ namespace RE
...
+	float NiPoint3::GetDistance(const NiPoint3& a_pt) const
+	{
+		float dx = a_pt.x - x;
+		float dy = a_pt.y - y;
+		float dz = a_pt.z - z;
+		return std::sqrtf(dx * dx + dy * dy + dz * dz);
+	}
+
+
+	float NiPoint3::GetSquaredDistance(const NiPoint3& a_pt) const
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.cpp:110]:
```suggestion
	float NiPoint3::GetSquaredDistance(const NiPoint3& a_pt) const noexcept
```

File: include/RE/N/NiPoint3.cpp [hunk 13]
```diff
@@ -98,6 +98,24 @@ namespace RE
 	}
 
 
+	float NiPoint3::GetDistance(const NiPoint3& a_pt) const
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/N/NiPoint3.cpp:101]:
```suggestion
	float NiPoint3::GetDistance(const NiPoint3& a_pt) const noexcept
```

File: include/RE/B/BSFurnitureMarkerNode.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "RE/B/BSTArray.h"
+#include "RE/N/NiExtraData.h"
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/B/BSFurnitureMarkerNode.h:4]:
```suggestion
#include "RE/B/BSTArray.h"
#include "RE/N/NiExtraData.h"
#include "RE/N/NiPoint3.h"
```

File: include/RE/B/BSFurnitureMarkerNode.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace RE
+{
+	class NiPoint3;
+
+
+	class FurniturePosition
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/B/BSFurnitureMarkerNode.h:11]:
```suggestion
	class BSFurnitureMarker
```

File: include/RE/B/BSFurnitureMarkerNode.h [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+	};
+
+
+	class BSFurnitureMarkerNode : public NiExtraData
+	{
+	public:
+		inline static constexpr auto RTTI = RTTI_BSFurnitureMarkerNode;
+		inline static constexpr auto NiRTTI = NiRTTI_BSFurnitureMarkerNode;
+
+		BSTArray<FurniturePosition> positions;	// 18
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/B/BSFurnitureMarkerNode.h:43]:
```suggestion
		BSTArray<BSFurnitureMarker> markers;	// 18
```

File: include/RE/G/GridArray.h [hunk 17]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "RE/N/NiPoint3.h"
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/G/GridArray.h:3]:
```suggestion
```

File: include/RE/B/BSFurnitureMarkerNode.h [hunk 18]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "RE/B/BSTArray.h"
+#include "RE/N/NiExtraData.h"
+
+namespace RE
+{
+	class NiPoint3;
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/B/BSFurnitureMarkerNode.h:8]:
```suggestion
```

File: include/RE/T/TES.h [hunk 19]
```diff
@@ -55,78 +72,87 @@ namespace RE
...
-		std::uint64_t	   unk288;	// 288
-		SystemEventAdapter unk290;	// 290
-		std::uint64_t	   unk2A0;	// 2A0
-		std::uint64_t	   unk2A8;	// 2A8
-		std::uint64_t	   unk2B0;	// 2B0
+		std::uint64_t					unk070;					 // 070
+		GridCellArray*					gridCellArray;			 // 078
+		NiNode*							objectLOD;				 // 080
+		NiNode*							landLOD;				 // 088
+		NiNode*							waterLOD;				 // 090
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:79]:
```suggestion
		GridCellArray*					gridCells;			 // 078
		NiNode*							objRoot;				 // 080
		NiNode*							lodLandRoot;				 // 088
		NiNode*							objLODWaterRoot;				 // 090
```

File: include/RE/T/TES.h [hunk 20]
```diff
@@ -55,78 +72,87 @@ namespace RE
...
+		BSTempNodeManager*				tempNodeManager;		 // 098
+		std::uint64_t					unk0A0;					 // 0A0
+		std::uint64_t					unk0A8;					 // 0A8
+		std::uint64_t					unk0B0;					 // 0B0
+		std::uint64_t					unk0B8;					 // 0B8
+		TESObjectCELL*					currentInteriorCell;	 // 0C0
+		TESObjectCELL**					interiorCellBuffer;		 // 0C8
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:86]:
```suggestion
		TESObjectCELL*					interiorCell;	 // 0C0
		TESObjectCELL**					interiorBuffer;		 // 0C8
```

File: include/RE/T/TES.h [hunk 21]
```diff
@@ -55,78 +72,87 @@ namespace RE
...
+		std::uint64_t					unk0E8;					 // 0E8
+		std::uint64_t					unk0F0;					 // 0F0
+		std::uint64_t					unk0F8;					 // 0F8
+		Sky*							sky;					 // 100
+		ImageSpaceModifierInstanceForm* currentImod;			 // 108
+		std::uint64_t					unk110;					 // 110
+		std::uint64_t					unk118;					 // 118
+		std::uint64_t					unk120;					 // 120
+		std::uint8_t					unk128;					 // 128
+		bool							lastToggleBordersState;	 // 129
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:99]:
```suggestion
		bool							showLANDborders;	 // 129
```

File: include/RE/T/TES.h [hunk 22]
```diff
@@ -55,78 +72,87 @@ namespace RE
...
+		std::uint8_t					unk12D;					 // 12D
+		std::uint8_t					unk12E;					 // 12E
+		std::uint8_t					unk12F;					 // 12F
+		std::uint16_t					unk130;					 // 130
+		std::uint8_t					unk132;					 // 132
+		std::uint8_t					unk133;					 // 133
+		std::uint32_t					unk134;					 // 134
+		std::uint64_t					unk138;					 // 138
+		TESWorldSpace*					worldSpace;				 // 140
+		BSSimpleList<NPCDeadCount*>		npcDeadCount;			 // 148
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:112]:
```suggestion
		BSSimpleList<BSTTuple<TESActorBase*, std::uint16_t>*>		deadCount;			 // 148
```

File: include/RE/T/TES.h [hunk 23]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "RE/B/BSTEvent.h"
+#include "RE/B/BSTList.h"
 #include "RE/I/ICellAttachDetachEventSource.h"
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:5]:
```suggestion
#include "RE/B/BSTEvent.h"
#include "RE/B/BSTList.h"
#include "RE/B/BSTTuple.h"
#include "RE/I/ICellAttachDetachEventSource.h"
#include "RE/N/NiSmartPointer.h"
```

File: include/RE/T/TES.h [hunk 24]
```diff
@@ -11,10 +12,17 @@ namespace RE
...
 	class BSSystemEvent;
-	struct PositionPlayerEvent;
+	class BSTempNodeManager;
+	class GridCellArray;
+	class ImageSpaceModifierInstanceForm;
+	class NiNode;
+	class Sky;
+	class TESNPC;
+	class TESObjectCELL;
+	class TESWorldSpace;
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:23]:
```suggestion
	class BSTempNodeManager;
	class GridCellArray;
	class ImageSpaceModifierInstance;
	class NiNode;
	class Sky;
	class TESActorBase;
	class TESObjectCELL;
	class TESWorldSpace;
```

File: include/RE/T/TES.h [hunk 25]
```diff
@@ -24,6 +32,15 @@ namespace RE
...
 		inline static constexpr auto RTTI = RTTI_TES;
 
+		struct NPCDeadCount
+		{
+			TESNPC*		  npc;		  // 00
+			std::uint16_t deadCount;  // 08
+			std::uint16_t pad0A;	  // 0A
+			std::uint32_t pad0C;	  // 0C
+		};
+		static_assert(sizeof(NPCDeadCount) == 0x10);
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:42]:
```suggestion
```

File: include/RE/T/TES.h [hunk 26]
```diff
@@ -55,78 +72,87 @@ namespace RE
...
+		TESObjectCELL**					interiorCellBuffer;		 // 0C8
+		void*							unk0D0;					 // 0D0
+		std::uint64_t					unk0D8;					 // 0D8
+		std::uint64_t					unk0E0;					 // 0E0
+		std::uint64_t					unk0E8;					 // 0E8
+		std::uint64_t					unk0F0;					 // 0F0
+		std::uint64_t					unk0F8;					 // 0F8
+		Sky*							sky;					 // 100
+		ImageSpaceModifierInstanceForm* currentImod;			 // 108
+		std::uint64_t					unk110;					 // 110
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:95]:
```suggestion
		BSSimpleList<NiPointer<ImageSpaceModifierInstance>> activeImageSpaceModifiers;			 // 108
```

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

File: include/RE/G/GridCellArray.h [hunk 27]
```diff
@@ -24,12 +24,11 @@ namespace RE
...
 		virtual void Unk_08(void) override;	 // 08
 
-		TESObjectCELL* GetCell(std::uint32_t a_x, std::uint32_t a_y)
+		[[nodiscard]] TESObjectCELL* GetCell(std::uint32_t a_x, std::uint32_t a_y) const noexcept
 		{
-			if (a_x >= length || a_y >= length) {
-				return nullptr;
-			}
-			return cells[(a_x * length) + a_y];
+			return (a_x >= length || a_y >= length) ?
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/G/GridCellArray.h:29]:
```suggestion
			return (a_x < length && a_y < length) ?
```

File: include/RE/T/TES.h [hunk 28]
```diff
@@ -72,87 +65,86 @@ namespace RE
...
-		std::uint64_t					unk278;					 // 278
-		std::uint64_t					unk280;					 // 280
-		std::uint64_t					unk288;					 // 288
-		SystemEventAdapter				unk290;					 // 290
-		std::uint64_t					unk2A0;					 // 2A0
-		std::uint64_t					unk2A8;					 // 2A8
-		std::uint64_t					unk2B0;					 // 2B0
+		std::uint64_t										  unk070;					  // 070
+		GridCellArray*										  gridCells;				  // 078
+		NiNode*												  objectRoot;				  // 080
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/T/TES.h:70]:
```suggestion
		NiNode*												  objRoot;				  // 080
```

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

**Pull request merged**